pkg/talos/nodeconfig.go

```diff
@@ -126,6 +138,106 @@ func applyNodeOverride(node *config.Node, cfg taloscfg.Provider) taloscfg.Provid
+	// changes.
+	renderedConfig := cfg.RawV1Alpha1()
+	unchangedConfig := renderedConfig.DeepCopy()
+
+	err := templateConfigField(node.NodeLabels, &renderedConfig.MachineConfig.MachineNodeLabels, unchangedConfig, "node labels")
+	if err != nil {
+		return nil, err
+	}
+
+	err = templateConfigField(node.NodeAnnotations, &renderedConfig.MachineConfig.MachineNodeAnnotations, unchangedConfig, "node annotations")
+	if err != nil {
+		return nil, err
+	}
+
+	return cfg, nil
+}
+
+// Given a map key pairs, template the values with the provided config, and apply the updated values to the destination map.
+// A message is written to the debug log with the field name and templated values.
+func templateConfigField(srcKeyPairs map[string]string, dstKeyPairs *map[string]string, cfg *v1alpha1.Config, fieldName string) error {
+	_, templateKeyPairs := splitTemplatedMapItems(srcKeyPairs)
+	if len(templateKeyPairs) == 0 {
+		return nil
+	}
+
+	renderedKeyPairs, err := templateMap(templateKeyPairs, cfg)
+	if err != nil {
+		return err
+	}
+
+	slog.Debug(fmt.Sprintf("adding %v to %s", renderedKeyPairs, fieldName))
+	for key := range renderedKeyPairs {
+		(*dstKeyPairs)[key] = renderedKeyPairs[key]
+	}
+
+	return nil
+}
+
+// Given a map `src`, return two maps where the first contains key-value pairs that are not
+// templates, and the latter containers only key-value pairs that are templates.
+func splitTemplatedMapItems(src map[string]string) (map[string]string, map[string]string) {
+	nonTemplateKeyPairs := make(map[string]string, len(src))
+	templateKeyPairs := make(map[string]string, len(src))
+
+	for key, value := range src {
+		if isTemplateString(value) {
+			templateKeyPairs[key] = value
+		} else {
+			nonTemplateKeyPairs[key] = value
+		}
+	}
+
+	return nonTemplateKeyPairs, templateKeyPairs
+}
+
+// Returns true if the provided string contains a template, false otherwise
+func isTemplateString(str string) bool {
+	return strings.Contains(str, "{{")
+}
+
+func templateMap(templateText map[string]string, data any) (map[string]string, error) {
+	// Track the new values separately so that the output is always deterministic.
+	// Otherwise, because map iteration order is random, multiple runs could
+	// produce different results in the case of referential values.
+	templatedText := make(map[string]string, len(templateText))
+	for key, templateValue := range templateText {
+		renderedValue, err := renderTemplate(templateValue, data)
+		if err != nil {
+			return nil, err
+		}
+
+		templatedText[key] = renderedValue
+	}
+
+	return templatedText, nil
+}
+
+// Takes in a Go template string and renders it.
+func renderTemplate(templateText string, data any) (string, error) {
+	builtTemplate, err := template.New("template").Funcs(sprig.FuncMap()).Parse(templateText)
+	if err != nil {
+		return "", err
+	}
+
+	buf := new(bytes.Buffer)
+	err = builtTemplate.Execute(buf, data)
+	if err != nil {
+		return "", err
+	}
+
+	return buf.String(), nil
+}
+
```

Comment:
It might be beneficial to move this logic (or most of it) to another package. What do you think?

Reply:
Yeah you're right.